### PR TITLE
remove redundant _has_run logic in val loop

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Fixes https://github.com/PyTorchLightning/pytorch-lightning/pull/8171#discussion_r660020944

Plus adds coverage for an untested `MisconfigurationException` in the `ModelCheckpoint` callback.

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/callbacks/model_checkpoint.py [hunk 1]
```diff
@@ -652,15 +652,11 @@ def _validate_monitor_key(self, trainer: 'pl.Trainer') -> None:
 
         # validate metric
         if self.monitor is not None and not self._is_valid_monitor_key(metrics):
-            m = (
+            raise MisconfigurationException(
```
Comment (tchaton) [inline pytorch_lightning/callbacks/model_checkpoint.py:655]: So would this raise an exception even if the val_loop hasn't run ?

File: tests/checkpointing/test_model_checkpoint.py [hunk 2]
```diff
@@ -1317,3 +1317,39 @@ def test_trainer_checkpoint_callback_bool(tmpdir):
...
+    )
+    model = LoggingModel()
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        logger=False,
+        max_steps=2,
+        num_sanity_val_steps=num_sanity_val_steps,
+        callbacks=[mc],
+    )
+    with pytest.raises(MisconfigurationException, match=r"ModelCheckpoint\(monitor='invalid'\) not found"):
```
Comment (tchaton) [inline tests/checkpointing/test_model_checkpoint.py:1350]: I am not sure to understand this test. It shouldn't raise an exception if validation hasn't been run.

Comment (awaelchli) [reply tests/checkpointing/test_model_checkpoint.py:1350]: this is a draft. 
I'm trying to undestand why _has_run was introduced.

File: pytorch_lightning/callbacks/model_checkpoint.py (hunk 1 above)
Comment (ananthsub) [reply pytorch_lightning/callbacks/model_checkpoint.py:627]: _validate_monitor_key would be called whenever the checkpoint is going to be saved. this could happen after training batches too, depending on the callback configuration. checking if the val loop has run, and if ithasn't, raising a warning vs if it has, raising an exception, silently hides this case where the monitor metric is misconfigured. 

i think its fair to enforce that if a metric key is provided, it must be present at the time we're saving a checkpoint
